♻️ Maintenance: Improve TypeScript support

segersniels opened the pull request:
## Description

After reading reports from people trying their hand on converting files to TS, I noticed the setup wasn't fully allowing conversion without causing a big waterfall of required changes. While it was already possible to import Flow files into TS, it apparently wasn't yet possible to import TS files into Flow without the need for a lot of changes. This PR basically attempts to fix whatever was needed so proper TS work can start again 🚀

### Changed

- Adjusted `babel` to properly handle all `.ts` files and not just the ones in `./src`
- Adjusted `tsconfig.json` to properly output ESM so it can be used alongside other ESM modules like `meow` and `execa`
- `withClient.ts` has been converted to TS 

### Fixed

- Jest now properly looks for `.ts` files and includes them in the test run
- `cli.spec.ts` has not been running since the original TS introduction and is now running again

Issue: https://github.com/carloscuesta/gitmoji-cli/issues/1110#issuecomment-1996843360

## Tests

- [x] All tests passed.

segersniels commented:
@carloscuesta I know the TS version of `cli.spec.js` hasn't been running since the refactor, but it seems that when I revert it back to the state it was before the refactor it doesn't even run. Did that test ever run? It seems like a bunch of stubs are missing to actually make everything work.

carloscuesta commented:
> @carloscuesta I know the TS version of `cli.spec.js` hasn't been running since the refactor, but it seems that when I revert it back to the state it was before the refactor it doesn't even run. Did that test ever run? It seems like a bunch of stubs are missing to actually make everything work. Not entirely familiar with jest as I usually work with mocha & chai so I might be missing something obvious.

Hey!

It's probably because of this:

https://github.com/carloscuesta/gitmoji-cli/blob/150a3505c9bd6d1c0012e7cc944b5e187a34a280/package.json#L97-L99

This is the pattern that jest uses to find which spec files to run, since the `cli.spec` was moved to `TS` we need to adjust this to:

```json
"testMatch": [
      "**/*.(spec).(js)",
      "**/*.(spec).(ts)"
    ],
```

So jest can find the `.spec.ts` files 👍🏼

segersniels commented:
> > @carloscuesta I know the TS version of `cli.spec.js` hasn't been running since the refactor, but it seems that when I revert it back to the state it was before the refactor it doesn't even run. Did that test ever run? It seems like a bunch of stubs are missing to actually make everything work. Not entirely familiar with jest as I usually work with mocha & chai so I might be missing something obvious.
> 
> Hey!
> 
> It's probably because of this:
> 
> https://github.com/carloscuesta/gitmoji-cli/blob/150a3505c9bd6d1c0012e7cc944b5e187a34a280/package.json#L97-L99
> 
> This is the pattern that jest uses to find which spec files to run, since the `cli.spec` was moved to `TS` we need to adjust this to:
> 
> ```json
> "testMatch": [
>       "**/*.(spec).(js)",
>       "**/*.(spec).(ts)"
>     ],
> ```
> 
> So jest can find the `.spec.ts` files 👍🏼

That was the original issue indeed, which we missed back when we refactored a couple of months ago. Test has been changed back to `cli.spec.js` in this PR and is running again, but failing on missing stubs. So I was wondering if it ever ran correctly or something broke with the jest mocks. You probably know more about jest than me

carloscuesta commented:
Yes, the test was running and passing successfully since it was added you can check it for example by going to the commit previous to the introduction of TS support and running the test:

```
git checkout c0b0ab8
yarn install
yarn test
```

Probably what happened is that as we missed including TS specs as part of that PR likely the test broke due to dependency updates that happened in the middle.

![CleanShot 2024-03-21 at 22 08 37@2x](https://github.com/carloscuesta/gitmoji-cli/assets/7629661/fd83bed8-57dc-4513-bed3-7558514420f8)

carloscuesta commented:
Actually the issue is that the mock was not properly implemented on that PR (not because any dependency update)

segersniels commented:
> Yes, the test was running and passing successfully from the beginning of it's inception you can check it for example by going to the commit previous to the introduction of TS support and running the test:
> 
> ```
> git checkout c0b0ab8
> yarn install
> yarn test
> ```
> 
> Actually the issue is that the mock was not properly implemented on that PR https://github.com/carloscuesta/gitmoji-cli/pull/1125/files#r1229243943
> 
> ![CleanShot 2024-03-21 at 22 08 37@2x](https://private-user-images.githubusercontent.com/7629661/315665604-fd83bed8-57dc-4513-bed3-7558514420f8.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTEwNTY4MjMsIm5iZiI6MTcxMTA1NjUyMywicGF0aCI6Ii83NjI5NjYxLzMxNTY2NTYwNC1mZDgzYmVkOC01N2RjLTQ1MTMtYmVkMy03NTU4NTE0NDIwZjgucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDMyMSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAzMjFUMjEyODQzWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YjJlMTg5MzYzNWVjM2UzMTUxNDViZGU4MDY1Nzg0NTk0ZjQ3YTdlNTViOWZmYWExZjQ4ZjY2ZjNkMGEwZTZkYSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.HevEGZwns_yDRGqxrFtTVYkeXb7vgZbFyBsePM11jHM)

I reverted all changes for that file in this PR but it's still not running it seems. So it should be identical to what it was before the refactor.

segersniels commented on package.json:
```diff
@@ -61,6 +61,7 @@
   "devDependencies": {
     "@babel/cli": "7.23.9",
     "@babel/core": "7.24.0",
+    "@babel/plugin-syntax-dynamic-import": "^7.8.3",
```
This one is apparently included in babel since `@babel/core@7.8.0`

carloscuesta commented:
It's tricky to exactly pin point what's the source of the issue now, as the spec file was not running since they refactor and several dependency updates were merged. What we certainly now is that before that specific commit it was passing as of this https://github.com/carloscuesta/gitmoji-cli/pull/1279#issuecomment-2013740537

So if you want to understand where the issue was exactly introduced (after reverting the TS PR) what I would suggest is doing a `bisect` https://git-scm.com/docs/git-bisect setting this hash as good `c0b0ab8` and the latest commit in master as bad

You'll have to revert the TS PR anyway bc the mock is not properly implemented

segersniels commented:
> > I reverted all changes for that file in this PR but it's still not running it seems. Which is weird to me because it should be identical to what it was before the refactor.
> 
> It's tricky to exactly pin point what's the source of the issue now, as the spec file was not running since the refactor and several dependency updates were merged. What we certainly know is that before that specific commit it was passing as of this [#1279 (comment)](https://github.com/carloscuesta/gitmoji-cli/pull/1279#issuecomment-2013740537)
> 
> So if you want to understand where the issue was exactly introduced (after reverting the TS PR) what I would suggest is doing a `bisect` https://git-scm.com/docs/git-bisect setting this hash as good [c0b0ab8](https://github.com/carloscuesta/gitmoji-cli/commit/c0b0ab8d9460400712718a49ff2191e4412c479a) and the latest commit in master as bad

My guess is the dynamic imports that we introduced as part of speed improvements.

carloscuesta commented on tsconfig.json:
```diff
@@ -1,8 +1,8 @@
     "target": "es6",
-    "module": "esnext",
-    "moduleResolution": "node",
+    "module": "node16",
+    "moduleResolution": "node16",
```
Why we need to change this? 🤔

segersniels replied:
Else TS complains that it can't find the `execa` and `meow` modules. See https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c

carloscuesta replied:
Thanks for the reference!

carloscuesta commented on package.json:
```diff
@@ -92,21 +92,27 @@
       "@commands/(.*)$": "<rootDir>/src/commands/$1",
-      "@constants/(.*)$": "<rootDir>/src/constants/$1"
+      "@constants/(.*)$": "<rootDir>/src/constants/$1",
+      "^(\\.\\.?\\/.+)\\.js$": "$1"
```
This is to allow importing files without specifying the `.js` extension right?

segersniels replied:
This is required to make sure `jest` can resolve the new `.js` suffix imports. If we don't specify it the tests will fail because it can't find the modules.

carloscuesta replied:
I see, so part of the transition to make this an ESM (since it's a requirement to specify extension on ESMs)
